pkg/clusterversion/cv.go

```diff
@@ -64,45 +64,60 @@ func (c *clusterVersionClient) GetClusterVersion() (*configv1.ClusterVersion, er
 	return cv, err
 }
 
-func (c *clusterVersionClient) EnsureDesiredVersion(uc *upgradev1alpha1.UpgradeConfig) (bool, error) {
+func (c *clusterVersionClient) EnsureDesiredConfig(uc *upgradev1alpha1.UpgradeConfig) (bool, error) {
 	clusterVersion, err := c.GetClusterVersion()
 	if err != nil {
 		return false, err
 	}
 
-	// Move the cluster to the same channel first
 	desired := uc.Spec.Desired
-	if clusterVersion.Spec.Channel != desired.Channel {
-		clusterVersion.Spec.Channel = desired.Channel
-		err = c.client.Update(context.TODO(), clusterVersion)
-		if err != nil {
-			return false, err
+
+	// Commence upgrade using version and channel
+	// Move the cluster to the same channel first
+	if len(desired.Channel) > 0 && len(desired.Version) > 0 {
+		if clusterVersion.Spec.Channel != desired.Channel {
+			clusterVersion.Spec.Channel = desired.Channel
+			err = c.client.Update(context.TODO(), clusterVersion)
+			if err != nil {
+				return false, err
+			}
+
+			// Retrieve the updated version
+			clusterVersion, err = c.GetClusterVersion()
+			if err != nil {
+				return false, err
+			}
+		}
+
+		// The CVO may need time sync the version before launching the upgrade
+		updateAvailable := false
+		for _, update := range clusterVersion.Status.AvailableUpdates {
+			if update.Version == desired.Version && update.Image != "" {
+				updateAvailable = true
+			}
+		}
+		if !updateAvailable {
+			return false, nil
 		}
 
-		// Retrieve the updated version
-		clusterVersion, err = c.GetClusterVersion()
+		clusterVersion.Spec.Overrides = []configv1.ComponentOverride{}
+		clusterVersion.Spec.DesiredUpdate = &configv1.Update{Version: uc.Spec.Desired.Version}
+		err = c.client.Update(context.TODO(), clusterVersion)
 		if err != nil {
 			return false, err
 		}
 	}
 
-	// The CVO may need time sync the version before launching the upgrade
-	updateAvailable := false
-	for _, update := range clusterVersion.Status.AvailableUpdates {
-		if update.Version == desired.Version && update.Image != "" {
-			updateAvailable = true
+	// Commence upgrade using Image
+	if len(desired.Image) > 0 {
+		if clusterVersion.Spec.DesiredUpdate.Image != desired.Image {
+			clusterVersion.Spec.DesiredUpdate.Image = desired.Image
```

Comment:
This needs to check that `clusterVersion.Spec.DesiredUpdate` is not nil (which will be the case on a fresh cluster), or else it will panic.
And it needs to initialize `DesiredUpdate` if it is nil, or else it will panic when setting `Image`.

Something like:
```suggestion
		if clusterVersion.Spec.DesiredUpdate == nil || (clusterVersion.Spec.DesiredUpdate.Image != desired.Image) {
			clusterVersion.Spec.DesiredUpdate = &configv1.Update{Image: desired.Image}
```

Reply:
done